cards.c

```diff
@@ -0,0 +1,329 @@
+	trQuestVarSet("cardUnits_" + index + "_Range",trQuestVarGet("card_" + proto + "_Range"));
+	trQuestVarSet("cardUnits_" + index + "_Keywords",trQuestVarGet("card_" + proto + "_Keywords"));
+	trStringQuestVarSet("cardUnits_" + index + "_Ability",trStringQuestVarGet("card_" + proto + "_Ability"));
+	trQuestVarSet("cardUnitsIndex", index + 1);
+	
+	DeploySober(protoName,vectorName);
+	trUnitChangeName("("+1*trQuestVarGet("card_" + proto + "_Cost")+") "+trStringQuestVarGet("card_" + proto + "_Name")+" <"+1*trQuestVarGet("card_" + proto + "_Speed")+">");
+}
+
+void CardLoad(bool firstBit = false, bool secondBit = false, int index = 0){
+	int copies = 0;
+	if(firstBit){
+		if(secondBit){
+			copies = 3;
+		} else {
+			copies = 1;
+		}
+	} else {
+		if(secondBit){
+			copies = 2;
+		}
+	}
+	if(copies>0){
+		trChatSend(0, "Player has " + copies + " copies of " + trStringQuestVarGet("card_" + 1*trQuestVarGet("cardProtos_"+index) + "_Name"));
+	}
+}
+
+rule initializeCards
+highFrequency
+active
+runImmediately
+{
+	for(p=1;<cNumberPlayers){
+		trForbidProtounit(p, "Archer Atlantean Hero");
+		trForbidProtounit(p, "Javelin Cavalry Hero");
+		trForbidProtounit(p, "Lancer Hero");
+		trForbidProtounit(p, "Maceman Hero");
+		trForbidProtounit(p, "Oracle Hero");
+		trForbidProtounit(p, "Royal Guard Hero");
+		trForbidProtounit(p, "Swordsman Hero");
+		trForbidProtounit(p, "Trident Soldier Hero");					
+	}
+
+	//Pick a card. Any card.
+	CardSetup("Militia", 				1, "Farmer", 			1, 3, 1, 0);
+	CardSetup("Maceman", 				1, "Squire", 			1, 2, 1, 0, Guard());
+	CardSetup("Ulfsark", 				1, "Wolfpelt", 			1, 2, 1, 0, Charge());
+	CardSetup("Slinger", 				1, "Gypsy", 			1, 1, 1, 1);
+	CardSetup("Toxotes", 				2, "Townguard Archer", 	2, 2, 1, 1);
+	CardSetup("Spearman", 				2, "Roadside Bandit", 	3, 2, 1, 0);
+	CardSetup("Anubite", 				2, "Dark Dog", 			2, 1, 2, 0, Charge());
+	CardSetup("Raiding Cavalry",		3, "Wild Horseman", 	2, 1, 3, 0);
+	CardSetup("Wadjet", 				3, "Noble Cobra", 		2, 3, 1, 2, Regenerate());
+	CardSetup("Ballista", 				4, "Giant Crossbow", 	3, 2, 1, 3);
+	CardSetup("Trident Soldier",		4, "Throne Shield", 	2, 6, 1, 0, Guard());
+	CardSetup("Avenger", 				5, "Avian Warrior", 	3, 5, 1, 0, Airdrop(Furious()));
+	CardSetup("Battle Boar", 			5, "Golden Hog", 		4, 4, 2, 0, Charge(Guard()));
+	CardSetup("Hero Greek Odysseus",	9, "Nickonhawk", 		4, 7, 1, 0, Guard(Ethereal(Regenerate())), "Listening to anime music and eating trash.");
+	CardSetup("Hoplite", 				9, "Zenophobia", 		7, 4, 1, 0, Charge(Armor(Deadly())), "Currently hard at work on Star Fighter Ace 3.");
+	CardSetup("Pharaoh of Osiris", 		9, "Yeebaagooon", 		9, 9, 9, 0, Airdrop(Ward(Lightning())), "Seraph addicted to dota and tea.");
+	
+	//Loading player collection
+	int cardIndex = 0;
+	for(i=0;<16){
+		int n = trGetScenarioUserData(i);
+		CardLoad((n<0), GetBit(n,0), cardIndex);
+		cardIndex = cardIndex + 1;
+		int j=1;
+		while(j<29){
+			CardLoad(GetBit(n,j), GetBit(n,j+1), cardIndex);
```

Comment:
This loop could be shortened to O(n) time if manual bit manipulation happens every loop